Locking rework

mahkoh opened the pull request:
This PR removes the editable_lock function.

This has several advantages:

- You no longer have to worry about calling editable_lock
- At least one deadlock has been removed by this change
- ~~The code is a about 20 lines smaler~~

To do this I've changed the way jobs modify the structures previously guarded by this mutex:

1. Jobs are scheduled as before
2. Once the jobs are done, they store the result in a list and notify the ui thread of this
3. The ui thread fetches the result and performs the necessary modifications

mahkoh commented on job.c:
```diff
@@ -311,74 +404,147 @@ void do_add_job(void *data)
+		for (size_t i = 0; i < d->used; i++)
+			track_info_unref(d->ti[i]);
+		free(d->ti);
+	}
```
this also fixes an unrelated memory leak

mahkoh commented:
I used a static analysis tool to make sure that editable_lock is only called from the places I fixed here. Unfortunately I missed one instance because it is called through a function pointer there. I've already written a patch and will push it tomorrow.

mahkoh commented:
This should be ready now.

flyingmutant commented on cmus.c:
```diff
@@ -57,33 +57,25 @@ int cmus_init(void)
 }
 
+static struct track_info *cmus_get_next_from_main_thread(void);
+
```
I think it is better to not have this forward declaration and simply use `cmus_get_next_track` here.

flyingmutant commented on cmus.c:
```diff
@@ -57,33 +57,25 @@ int cmus_init(void)
-		}
-	}
-	editable_unlock();
+	info = cmus_get_next_from_main_thread();
```
This can be merged with `info` declaration above.

flyingmutant commented on cmus.c:
```diff
@@ -164,15 +154,17 @@ enum file_type cmus_detect_ft(const char *name, char **ret)
 
-void cmus_add(add_ti_cb add, const char *name, enum file_type ft, int jt, int force)
+void cmus_add(add_ti_cb add, const char *name, enum file_type ft, int jt,
+		int force)
```
I think it is better to try to keep things like this as single line – even if it crosses some (arbitrary) line length limit.

mahkoh replied:
Some line length limit should be used. If you think 80 is too little, what about 90?

flyingmutant replied:
90 is probably OK.

flyingmutant commented on cmus.c:
```diff
@@ -388,3 +376,72 @@ int cmus_playlist_for_each(const char *buf, int size, int reverse,
 	return 0;
 }
+
+/* multit-hreaded next track requests */
```
Typo.

flyingmutant commented on cmus.c:
```diff
@@ -388,3 +376,72 @@ int cmus_playlist_for_each(const char *buf, int size, int reverse,
+static pthread_mutex_t cmus_next_file_mutex = CMUS_MUTEX_INITIALIZER;
+static pthread_cond_t cmus_next_file_cond = CMUS_COND_INITIALIZER;
+static int cmus_next_file_provided;
+static struct track_info *cmus_next_file;
```
I'd prefer to keep globals at the top of the file.

flyingmutant commented on job.c:
```diff
@@ -46,27 +47,116 @@
+static int job_fd_priv;
+
+static LIST_HEAD(job_result_head);
+static pthread_mutex_t job_mutex = CMUS_MUTEX_INITIALIZER;
```
Please keep globals at the top of the file.

flyingmutant commented on cmus.c:
```diff
@@ -388,3 +376,72 @@ int cmus_playlist_for_each(const char *buf, int size, int reverse,
+	cmus_mutex_lock(&mutex);
+
+	char buf = 0;
+	write(cmus_next_track_request_fd_priv, &buf, 1);
```
I'd prefer to have some `notify`-style wrappers around pipe-related read/write calls – it would make the code much more readable.

mahkoh commented:
I've merged the other locking PR into this. All commits are now self-contained, compile, and can be reviewed individually.

mahkoh commented:
I've addressed the style issues and fixed another bug. Namely https://github.com/cmus/cmus/pull/472/commits/b9c641c658af7510426d23a7c872eb09cc67f274

The rest of the commit history is unchanged.

flyingmutant commented:
Thanks *a lot* for cleanly separated commits — this made reviewing the PR much easier.
